### DISCUSSION TITLE
Hiding sensitive data during an API request

### DISCUSSION THREAD

**Discussion by alanakra** [vercel/next.js#28284] [Help] [answered]
I'm currently working on a form with Next.js whose data is sent to a Trello board via the latter's API.
<img width="739" alt="Capture d’écran 2021-08-19 à 10 30 24" src="https://user-images.githubusercontent.com/50582251/130041684-5a2fba5f-a632-46d1-a7c9-ecad457c3045.png">

This one works well, I have my data sent and I have a response, however I use an API key and a token that even stored on a .env file are visible on the HTTP headers which poses a security problem.
<img width="1440" alt="Capture d’écran 2021-08-19 à 09 40 46" src="https://user-images.githubusercontent.com/50582251/130041592-ff8ef012-09b8-49c7-8d3c-55576ea9afbf.png">
The other concern is that I am using Vercel for hosting which is serverless and cannot send information. So I wanted to know if you know of a cloud service that can handle this. 

Hiding sensitive data during an API request from a serverless app

Update the question so it can be answered with facts and citations. This will help others answer the question. You can edit the question or post a new one.

**Comment by logantai24** [comment, marked as answer]
@alanakra instead of directly sending out sensitive data (API token) in frontend, you might consider using the [api routes](https://nextjs.org/docs/api-routes/introduction) functionality of Nextjs? 

**Nextjs App (frontend)** 
Sends a POST request to an api route you created, which includes the form data. 

**API (Vercel Serverless Functions)**
Retrieve the form data, send a POST request to the trello API and write that form data to Trello. Sensitive data is store in env which only exposes to this API. 

By doing so, the sensitive data will not be exposed in browser as all the actual _send form data to trello API_ logic is done within the API endpoint you created in the pages/api/ folder.

**Reply by alanakra** [reply]
I'll try this way. Thanks

**Reply by elouisramsey** [reply]
@alanakra were you able to achieve this?

**Reply by alanakra** [reply]
@elouisramsey Hello,
This was a project during my end of study internship in a company. I was able to do it, but I don't have access to the source code anymore.

**Reply by afaqahmedkhan** [reply]
@elouisramsey were you able to achieve this?

**Comment by elouisramsey** [comment]
Hello, ended up encrypting and decrypting the data on the FE and BE, that
seems to work for now
